openedx_learning/apps/authoring/units/api.py

```diff
@@ -0,0 +1,300 @@
+"""Units API.
+
+This module provides functions to manage units.
+"""
+from dataclasses import dataclass
+from datetime import datetime
+
+from django.db.transaction import atomic
+
+from openedx_learning.apps.authoring.components.models import Component, ComponentVersion
+
+from ..containers import api as container_api
+from ..publishing.api import get_published_version_as_of
+from .models import Unit, UnitVersion
+
+# 🛑 UNSTABLE: All APIs related to containers are unstable until we've figured
+#              out our approach to dynamic content (randomized, A/B tests, etc.)
+__all__ = [
+    "create_unit",
+    "create_unit_version",
+    "create_next_unit_version",
+    "create_unit_and_version",
+    "get_unit",
+    "get_unit_version",
+    "get_latest_unit_version",
+    "UnitListEntry",
+    "get_components_in_draft_unit",
+    "get_components_in_published_unit",
+    "get_components_in_published_unit_as_of",
+]
+
+
+def create_unit(
+    learning_package_id: int, key: str, created: datetime, created_by: int | None
+) -> Unit:
+    """
+    [ 🛑 UNSTABLE ] Create a new unit.
+
+    Args:
+        learning_package_id: The learning package ID.
+        key: The key.
+        created: The creation date.
+        created_by: The user who created the unit.
+    """
+    with atomic():
+        container = container_api.create_container(
+            learning_package_id, key, created, created_by
+        )
+        unit = Unit.objects.create(
+            container=container,
+            publishable_entity=container.publishable_entity,
+        )
+    return unit
+
+
+def create_unit_version(
+    unit: Unit,
+    version_num: int,
+    *,
+    title: str,
+    publishable_entities_pks: list[int],
+    entity_version_pks: list[int | None],
+    created: datetime,
+    created_by: int | None = None,
+) -> UnitVersion:
+    """
+    [ 🛑 UNSTABLE ] Create a new unit version.
+
+    Args:
+        unit_pk: The unit ID.
+        version_num: The version number.
+        title: The title.
+        publishable_entities_pk: The publishable entities.
+        entity: The entity.
+        created: The creation date.
+        created_by: The user who created the unit.
+    """
+    with atomic():
+        container_version = container_api.create_container_version(
+            unit.container.pk,
+            version_num,
+            title=title,
+            publishable_entities_pks=publishable_entities_pks,
+            entity_version_pks=entity_version_pks,
+            created=created,
+            created_by=created_by,
+        )
+        unit_version = UnitVersion.objects.create(
+            unit=unit,
+            container_version=container_version,
+            publishable_entity_version=container_version.publishable_entity_version,
+        )
+    return unit_version
+
+
+def create_next_unit_version(
+    unit: Unit,
+    *,
+    title: str | None = None,
+    components: list[Component | ComponentVersion] | None = None,
```

Comment:
Something else that I wanted to call out is that many Learning Core APIs take integer IDs as their parameters but in several of these new APIs I'm taking `Unit` and `Component | ComponentVersion` objects. I think this makes the API much nicer (we can use types to distinguish between `Component` and `ComponentVersion` to determine if things should be pinned or not) and more efficient, but arguably it could be more consistent to only take integer IDs.

Reply:
@ormsbee could you weigh in please?

Reply:
I think it was mostly that there were places where I could get by with just an ID, like `create_component_version`, and then I kept things the same for consistency.

We discussed this a bit back in https://github.com/openedx/openedx-learning/issues/34. My last comment at the time was:

> For now, the API mostly uses primary keys internally, except some functions that are explicitly labeled as using keys. I'm closing this for now, with the understanding that we might revisit it if when/if the current conventions become a problem.

So yeah, I'm not wedded to it.